EE: Remote Repositories

himdel opened the pull request:
Fixes [AAH-437](https://issues.redhat.com/browse/AAH-437) (with #873)
API: https://github.com/ansible/galaxy_ng/pull/935 (merged), https://github.com/ansible/galaxy_ng/pull/985 (merged), https://github.com/ansible/galaxy_ng/pull/988
Design https://marvelapp.com/prototype/27ee1cd6/screen/82150243 
Soft-depends on #873 for Remote Registry crud (merged)

This allows adding, editing and syncing remote repositories in Execution Environments. These new remote container repos act like local ones, except only local ones can have tags managed while only remote ones can sync, and show sync status.

* EE list:
  * [x] Add button to sync EE if it is a remote EE.
  * [x] Add option to edit an EE’s remote.
  * [x] Add button to create new remote EE.
  * [x] Add if container registry is remote or not

![20210922234654](https://user-images.githubusercontent.com/289743/134436473-c1f8938a-971d-4384-938c-3b71e2bf9446.png)

* EE detail:
  * [x] last updated from registry
  * [x] edit merge
  * [x] disallow editing tags on remotes
  * [x] dropdown with sync&edit(&use)

![20210924172413](https://user-images.githubusercontent.com/289743/134716058-03a0dd95-61d7-4498-ab0c-a371a44cd1d3.png)

* the form itself:
  * [x] Create new remote repository: Allows the user to specify an upstream image name, registry and image name. Creates a new remote execution environment that can be synced
  * [x] Edit remote form. Allows updating a remotes upstream name and registry BUT NOT THE REMOTE’S NAME.
  * [x] rename from Edit repository to Edit execution environment

| ![20210924011502](https://user-images.githubusercontent.com/289743/134603644-10f5add9-bca6-41b1-847a-5117328abf05.png) | ![20210924011617](https://user-images.githubusercontent.com/289743/134603729-2858bce3-3b8d-4516-bc66-96a5a397d9e2.png) |
|-|-|

himdel commented:
(Currently this allows Add/Edit of remote repos from the EE list screen, except for groups&description, so I mostly just need to merge the 2 `onSave` bits to do both in both places)

himdel commented:
Testing:

registry url: `https://registry.hub.docker.com`
remote upstream_name: `library/alpine` (`_/alpine` on dockerhub becomes `library/alpine`)

ZitaNemeckova reviewed (approved):
LGTM :+1:

newswangerd commented:
Need to add the "create ee" button to the empty state page when there are no execution enviromments. Right now it's impossible to add a remote ee if the registry is empty.
![Screen Shot 2021-09-24 at 3 07 34 PM](https://user-images.githubusercontent.com/6063371/134727641-71822e18-7066-4e02-ad6d-7974cedbc1ee.png)

newswangerd commented on src/containers/execution-environment-list/execution_environment_list.tsx:
```diff
@@ -338,13 +386,68 @@ class ExecutionEnvironmentList extends React.Component<
+    const { name, namespace, description, pulp } = itemToEdit;
+    const { pulp_id, registry, upstream_name, include_tags, exclude_tags } =
+      pulp?.repository?.remote || {};
+    const remote = pulp?.repository ? !!pulp?.repository?.remote : true; // add only supports remote
```
Don't use `object?.` It's not supported on older browsers and we don't have a polyfill.

himdel replied:
Looks like `?.` gets compiled by babel, `Array.prototype.at` was the problematic one :)

newswangerd replied:
my bad :)

newswangerd commented on src/components/execution-environment/repository-form.tsx:
```diff
@@ -114,4 +340,120 @@ export class RepositoryForm extends React.Component<IProps, IState> {
+    } as any);
+  }
+
+  private onSave() {
```
I don't know where this is happening, but if the tasking system is busy (if a sync is happening) and you try to edit the description on an repository, the page spams the task endpoint for updates on the distribution task. If you navigate away from the page while the form is loading, the UI seems to keep spamming the API forever.

You need to rate limit the requests for updates from the task to 1 every 5 seconds or so and make sure to garbage collect the request if the component is unmounted.

himdel replied:
Updated to use a shared `waitForTask` helper & set the limit to 5s.
(still looking into making sure it ends at some point)

newswangerd reviewed (changes requested):
Getting an error on the images page after performing a sync

```
Uncaught TypeError: selectedImage is undefined
    renderImages execution_environment_detail_images.tsx:186
    render execution_environment_detail_images.tsx:114
    React 11
    unstable_runWithPriority scheduler.development.js:653
    React 6
    queryImages execution_environment_detail_images.tsx:436
```

himdel commented:
Updated...

`waitForTask` is now shared, and bails after 10 attempts
`selectedImage` being undefined is ok after deletion, handling the error
and added the Add EE button to empty state :)
